src/mlpack/core/data/dict_encoding.hpp

```diff
@@ -0,0 +1,188 @@
+ * @file dict_encoding.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace data {
+/**
+ * A simple Dictionary Enocding class
+ */
+class DicitonaryEncoding
+{
+ public:
+  /**
+  * Default constructor 
+  */
+  DicitonaryEncoding();
+
+  /**
+  * Default Destructor 
+  */
+  ~DicitonaryEncoding();
+
+  /**
+  * A function to create map using a corpus
+  * 
+  * @param strings Corpus of text to encode
+  * @param deliminator Delimiator use to split the corpus
+  */
+  void CreateMap(std::string strings, char deliminator = ' ');
+
+  /**
+  * Overlaoded function to split into characters
+  * Since Boost library doesn't have function to split into characters
+  *
+  * @param strings Corpus of text to encode
+  */
+  void CreateCharMap(std::string strings);
+
+  /**
+  * A function to reset the mapping that is clear all the encodings
+  */
+  void Reset();
+
+  /**
+  * A function to encode strings into characters
+  *
+  * for example 
+  * GATTACA into seven different values
+  * The above will be encode to 1233242
+  *
+  * @param strings Vector of strings
+  * @param output Output Matrix to store encoded results
+  * @param mapping Mappings use to encode input
+  *    
+  */
+  template<typename eT>
+  void CharEncode(const std::vector<std::string>& strings,
+              arma::Mat<eT>& output,
+              std::unordered_map<std::string, size_t>& mapping);
+  /**
+  * Overloaded function for a user to avoid passing of mappings
+  * Default Data member would be used for mapping
+  *
+  * @param strings Vector of strings
+  * @param output Output Matrix to store encoded results
+  *
+  */
+  template<typename eT>
+  void CharEncode(const std::vector<std::string>& strings,
+                  arma::Mat<eT>& output);
+
+  /**
+  * A fucntion to encode given array of strings using a particular deliminator
+  *
+  * For example 
+  * Vector is :
+  * [hello@wow, wow@hello@good] would be encoded using '@' as deliminator as 
+  * [1 2 0, 2 1 3] 
+  * The function paddes 0 to maintain same sizes across all the rows
+  *
+  * @param strings Vector of strings
+  * @param output Output Matrix to store encoded results
+  * @param mapping Mapping use to encode the input
+  * @param deliminator Delimnator used to split the strings
+  */
+  template<typename eT>
+  void DictEncode(const std::vector<std::string>& strings,
+                  arma::Mat<eT>& output,
+                  std::unordered_map<std::string, size_t>& mapping,
+                  const char deliminator = ' ');
```

Comment:
I think you implemented a special case. Thus, the delimiter can also be one character of many or even a string. Actually, we can't divine all possible cases. I suggest to add a template parameter e.g. `Tokenizer` which should have an operator `()` that accepts a string or `string_view` and returns a token. Moreover, we should implement some standard tokenizers.
P.S. In that case you needn't 2 different functions like `CharEncode` and `DictEncode`.

Reply:
In addition: maybe it is a good idea to add another abstraction for the purpose of cleaning data. For example we could introduce a template parameter `CleaningPolicy` as well as `Tokenizer`.

Reply:
> I suggest to call it "delimiter":).

changed it to delimiter

Reply:
> I think you implemented a special case. Thus, the delimiter can also be one character of many or even a string. Actually, we can't divine all possible cases. I suggest to add a template parameter e.g. `Tokenizer` which should have an operator `()` that accepts a string or `string_view` and returns a token. Moreover, we should implement some standard tokenizers.
> P.S. In that case you needn't 2 different functions like `CharEncode` and `DictEncode`.

Implemented a Tokenizer class, also provided a interface which i guess would be easy for user since a user may not want to provide a custom tokenization everytime.

Reply:
> In addition: maybe it is a good idea to add another abstraction for the purpose of cleaning data. For example we could introduce a template parameter `CleaningPolicy` as well as `Tokenizer`.

I have started working on algorithms which is required to preprocess text data, such as removing punctuation, stop words, and also changing everything to lower case. I think we can provide a separate class for it .